### PR TITLE
NT-1951: UI/UX – Error loading initial comments

### PR DESCRIPTION
**Pull request by sunday-okpoluaefe**
# 📲 What

UI/UX – Error loading initial comments

# 🤔 Why

- Because every now and then things will fail, we need to handle problems loading comments.

# 🛠 How

- Created custom error layout item
- Added custom item to viewholder
- Added Initial Load Error observable
- Subscribed to initialCommentLoadError
- Added unit tests for initial comment load error

# 👀 See

https://user-images.githubusercontent.com/63934292/121552959-c8a90000-ca08-11eb-93b9-8a6311506cb4.mp4

| Before 🐛 | After 🦋 |
| --- | --- |
|  |  |

# 📋 QA

- Select any project with comments
- Turn off network connect
- Select comments, you should see an error page on failure

# Story 📖

https://kickstarter.atlassian.net/browse/NT-1951

### CODE REVIEW

File: app/src/main/java/com/kickstarter/libs/utils/extensions/BoolenExt.kt [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package com.kickstarter.libs.utils.extensions
+
+import android.view.View
+
+fun Boolean.toVisibility(): Int {
```
Comment (Arkariang) [inline app/src/main/java/com/kickstarter/libs/utils/extensions/BoolenExt.kt:5]: love it! ❤️

File: app/src/main/java/com/kickstarter/ui/adapters/CommentsAdapter.kt [hunk 2]
```diff
@@ -4,27 +4,62 @@ import android.view.LayoutInflater
...
-    fun takeData(comments: List<CommentCardData>) {
+    private fun resetList() {
         clearSections()
-        addSection(comments)
+        insertSection(SECTION_INITIAL_LOAD_ERROR, emptyList<Boolean>())
+        insertSection(SECTION_COMMENT_CARD, emptyList<CommentCardData>())
+    }
+
+    fun takeData(comments: List<CommentCardData>) {
+        resetList()
```
Comment (Arkariang) [inline app/src/main/java/com/kickstarter/ui/adapters/CommentsAdapter.kt:43]: Do not call resetList(), that should be only initialization step

File: app/src/main/java/com/kickstarter/ui/adapters/CommentsAdapter.kt [hunk 3]
```diff
@@ -4,27 +4,62 @@ import android.view.LayoutInflater
...
+        setSection(SECTION_COMMENT_CARD, comments)
+        submitList(items())
+    }
+
+    fun insertPageError() {
+        resetList()
```
Comment (Arkariang) [inline app/src/main/java/com/kickstarter/ui/adapters/CommentsAdapter.kt:49]: same as before

File: app/src/main/java/com/kickstarter/ui/adapters/CommentsAdapter.kt [hunk 4]
```diff
@@ -4,27 +4,62 @@ import android.view.LayoutInflater
...
+        setSection(SECTION_INITIAL_LOAD_ERROR, listOf(true))
         submitList(items())
     }
 
     override fun viewHolder(@LayoutRes layout: Int, viewGroup: ViewGroup): KSViewHolder {
-        return CommentCardViewHolder(ItemCommentCardBinding.inflate(LayoutInflater.from(viewGroup.context), viewGroup, false), delegate)
+        return when (layout) {
+            R.layout.item_comment_card -> CommentCardViewHolder(ItemCommentCardBinding.inflate(LayoutInflater.from(viewGroup.context), viewGroup, false), delegate)
+            R.layout.comment_initial_load_error_layout -> InitialCommentLoadErrorViewHolder(
```
Comment (Arkariang) [inline app/src/main/java/com/kickstarter/ui/adapters/CommentsAdapter.kt:57]: I think you can use delete the `InitialCommentLoadErrorViewHolder` and reuse the already existing `EmptyViewHolder` just inflate the UI you created

File: app/src/main/java/com/kickstarter/viewmodels/CommentsViewModel.kt [hunk 5]
```diff
@@ -184,20 +191,38 @@ interface CommentsViewModel {
...
-                    it.localizedMessage
-                    Timber.d("************ On initializing error")
-                    this.initialError.onNext(it)
+                    this.initialError.onNext(it.first)
+                    this.isLoadingMoreItems.onNext(false)
                 }
 
-            // TODO showcasing pagination error subscriptionto be completed on : https://kickstarter.atlassian.net/browse/NT-2019
-            this.paginationError
+            // TODO showcasing pagination error subscription to be completed on : https://kickstarter.atlassian.net/browse/NT-2019
```
Comment (Arkariang) [inline app/src/main/java/com/kickstarter/viewmodels/CommentsViewModel.kt:205]: you can remove the TODO as the work is already done:)

File: app/src/main/java/com/kickstarter/viewmodels/CommentsViewModel.kt [hunk 6]
```diff
@@ -184,20 +191,38 @@ interface CommentsViewModel {
...
+            this.internalError
+                .filter { this.lastCommentCursor != null }
+                .compose(bindToLifecycle())
                 .subscribe {
-                    it.localizedMessage
-                    Timber.d("************ On pagination error")
+                    this.paginationError.onNext(it)
+                    this.isLoadingMoreItems.onNext(false)
```
Comment (Arkariang) [inline app/src/main/java/com/kickstarter/viewmodels/CommentsViewModel.kt:211]: Move this to the `OnError` for the network call

File: app/src/main/java/com/kickstarter/ui/viewholders/InitialCommentLoadErrorViewHolder.kt [hunk 7]
```diff
@@ -0,0 +1,11 @@
+package com.kickstarter.ui.viewholders
```
Comment (Arkariang) [inline app/src/main/java/com/kickstarter/ui/viewholders/InitialCommentLoadErrorViewHolder.kt:1]: This ViewHolder can be deleted and use instead `EmptyViewHolder` already existing in the app

File: app/src/main/res/values/strings.xml [hunk 8]
```diff
@@ -84,6 +84,7 @@
   <string name="leave_comment">Leave a comment…</string>
   <string name="only_backers_can_leave_comments">Only backers can leave comments</string>
   <!-- comment card -->
+  <string name="something_went_wrong">Something went wrong. Pull \ndown to reload this page.</string>
```
Comment (Arkariang) [inline app/src/main/res/values/strings.xml:87]: I think there should already be a translated string in place already, as the translation pull request was merged this morning

**Review by Arkariang — APPROVED** [review]
I left a few comments , nothing serious just small tweecks. 
Once those comments are addressed feel free to merge, overall it looks great :) 👍

File: app/src/main/java/com/kickstarter/ui/adapters/CommentsAdapter.kt (hunk 4 above)
Comment (sunday-okpoluaefe) [reply app/src/main/java/com/kickstarter/ui/adapters/CommentsAdapter.kt:57]: yeah. done

File: app/src/main/java/com/kickstarter/ui/adapters/CommentsAdapter.kt (hunk 3 above)
Comment (sunday-okpoluaefe) [reply app/src/main/java/com/kickstarter/ui/adapters/CommentsAdapter.kt:49]: Removing ``` resetList ``` causes some weird behavior. The previous error page persists even after data arrives

File: app/src/main/java/com/kickstarter/viewmodels/CommentsViewModel.kt (hunk 5 above)
Comment (sunday-okpoluaefe) [reply app/src/main/java/com/kickstarter/viewmodels/CommentsViewModel.kt]: done

File: app/src/main/java/com/kickstarter/ui/adapters/CommentsAdapter.kt (hunk 3 above)
Comment (Arkariang) [reply app/src/main/java/com/kickstarter/ui/adapters/CommentsAdapter.kt:49]: fyi: you need to clean specific sections not all of them. Meaning if you need to show the section "SECTION_INITIAL_LOAD_ERROR" send an empty list to the section "SECTION_COMMENT_CARD" and viceversa. For now it's not a problem but I'm gonna have to change it either way on my open PR(https://github.com/kickstarter/android-oss/pull/1284) as I need to present there 2 sections at the same time

File: app/src/main/res/values/strings.xml (hunk 8 above)
Comment (Arkariang) [reply app/src/main/res/values/strings.xml:87]: forget about this one, it seems still some translations not ready yet

**Pull request merged**